### ISSUE TITLE
vim.lsp.diagnostic.goto_next/goto_prev jumps to floating window

### ISSUE THREAD

**Issue by yujinyuz** [neovim/neovim#15122]
- `nvim --version`: 0.5.0
- Operating system/version: macOS
- Terminal name/version: alacritty
- `$TERM`: tmux-256color

### Steps to reproduce using `nvim -u NORC`

1. Install `pyright` (https://github.com/microsoft/pyright)
2. Setup LSP + `pyright` using settings from https://github.com/neovim/nvim-lspconfig
3. Open a new Python file, create one error on a single line
4. Add a mapping for goto_next and goto_prev
5. Run `:lua vim.lsp.diagnostic.goto_next()` and observe that cursor jumps to the floating diagnostic

### Actual behaviour

[![asciicast](https://asciinema.org/a/zxCC9IM8wNcFzSyHN3OnH94mZ.svg)](https://asciinema.org/a/zxCC9IM8wNcFzSyHN3OnH94mZ)

### Expected behaviour

It should not jump to floating terminal

**Comment by clason** [comment]
That's not the default behaviour, you must be using some plugins interacting with the LSP that behave wrongly.

**Comment by yujinyuz** [comment]
> That's not the default behaviour, you must be using some plugins interacting with the LSP that behave wrongly.

It happens when there is only one diagnostic available. But if there are more than 1 problems in a file, it works as expected

**Comment by clason** [comment]
Then I'm confused: Your reproducing steps say
> Run :lua vim.lsp.diagnostic.goto_next() and observe that the cursor jumps between lines, but not between diagnostics on the same line.

which implies multiple problems?

**Comment by yujinyuz** [comment]
> Then I'm confused: Your reproducing steps say
> 
> > Run :lua vim.lsp.diagnostic.goto_next() and observe that the cursor jumps between lines, but not between diagnostics on the same line.
> 
> which implies multiple problems?

apologies for the confusion. I've updated the steps

**Comment by clason** [comment]
Then this _is_ in fact intended behavior -- all default LSP handlers behave in the same way: if you trigger a request that opens a float twice without moving, you will enter that float (so you can scroll or yank). 

It's a bit unfortunate in this situation since you didn't know that you wouldn't be moving, but it is done for consistency.

You can always override the handler to implement your own behavior.

**Comment by yujinyuz** [comment]
@clason Thanks! That explanation makes sense. My expected behavior is just that it should do nothing and remain the same.

How can I override the handler or where can I read about it?

I've also tried lspsaga and it seems that it works the way I want

**Comment by clason** [comment]
I think simply `goto_next { focusable = false }` should work.

(And, yes, lspsaga has its own handlers.)

**Comment by yujinyuz** [comment]
Thanks!

For anyone interested, this is what I did:

```
  vim.api.nvim_buf_set_keymap(bufnr, 'n', ']w', [[<cmd>lua vim.lsp.diagnostic.goto_next({popup_opts = {focusable = false }})<cr>]], {noremap = true, silent = true})
  vim.api.nvim_buf_set_keymap(bufnr, 'n', '[w', [[<cmd>lua vim.lsp.diagnostic.goto_prev({popup_opts = {focusable = false }})<cr>]], {noremap = true, silent = true})

```

**Comment by fsouza** [comment]
> It's a bit unfortunate in this situation since you didn't know that you wouldn't be moving, but it is done for consistency.

TL;DR I think the default should be changed in goto_next and goto_prev

Now, this is a rant, but there's a difference between "side-effect caused by code reuse" and "consistency" 🙈 This is _not_ consistent at all because it only happens when there's only one diagnostic, and I'd love to find anyone who thinks that behavior is intuitive/expected 😬 

Seeing all LSP pop-ups as a single thing is a code reuse/maintainer point of view, not a user point of view. The fact that the code to manage popups used for signatureHelp, hover and goto_next is the same is an implementation detail, not a consistent experience for users.
